source/opt/inline_pass.cpp

```diff
@@ -0,0 +1,430 @@
+              }
+            });
+            // map and reset result id
+            const uint32_t rid = spv_inst->result_id();
+            if (rid != 0) {
+              const uint32_t nid = this->TakeNextId();
+              callee2caller[rid] = nid;
+              spv_inst->SetResultId(nid);
+            }
+            bp->AddInstruction(std::move(spv_inst));
+          } break;
+        }
+      });
+}
+
+bool InlinePass::Inline(ir::Function* func) {
+  bool modified = false;
+  for (auto bi = func->begin(); bi != func->end(); bi++) {
+    for (auto ii = bi->begin(); ii != bi->end();) {
+      if (ii->opcode() == SpvOp::SpvOpFunctionCall) {
+        // Inline call
+        std::vector<std::unique_ptr<ir::BasicBlock>> newBlocks;
+        std::vector<std::unique_ptr<ir::Instruction>> newVars;
+        GenInlineCode(&newBlocks, &newVars, ii, bi);
+        // update block map given replacement blocks
+        for (auto& blk : newBlocks) {
+          id2block_[blk->GetLabelId()] = &*blk;
+        }
+        // update phi functions in succesor blocks if call block
+        // is replaced with more than one block
+        if (newBlocks.size() > 1) {
+          const auto firstBlk = newBlocks.begin();
+          const auto lastBlk = newBlocks.end() - 1;
+          const uint32_t firstId = (*firstBlk)->GetLabelId();
+          const uint32_t lastId = (*lastBlk)->GetLabelId();
+          (*lastBlk)->ForEachSucc([&firstId, &lastId, this](uint32_t succ) {
+            ir::BasicBlock* sbp = this->id2block_[succ];
+            sbp->ForEachPhiInst([&firstId, &lastId](ir::Instruction* phi) {
+              phi->ForEachInId([&firstId, &lastId](uint32_t* id) {
+                if (*id == firstId) *id = lastId;
+              });
+            });
+          });
+        }
+        // replace old calling block with new block(s)
+        bi = bi.Erase();
+        bi = bi.InsertBefore(&newBlocks);
+        // insert new function variables
+        if (newVars.size() > 0) {
+          auto vbi = func->begin();
+          auto vii = vbi->begin();
+          vii.InsertBefore(&newVars);
+        }
+        // restart inlining at beginning of calling block
+        ii = bi->begin();
+        modified = true;
+      } else {
+        ii++;
+      }
+    }
+  }
+  return modified;
+}
+
+void InlinePass::Initialize(ir::Module* module) {
+  def_use_mgr_.reset(new analysis::DefUseManager(consumer(), module));
+
+  // Initialize next unused Id
+  nextId_ = 0;
+  for (const auto& id_def : def_use_mgr_->id_to_defs()) {
+    nextId_ = std::max(nextId_, id_def.first);
+  }
+  nextId_++;
+
+  module_ = module;
+
+  // initialize function and block maps
+  id2function_.clear();
+  id2block_.clear();
+  for (auto& fn : *module_) {
+    id2function_[fn.GetResultId()] = &fn;
+    for (auto& blk : fn) {
+      id2block_[blk.GetLabelId()] = &blk;
+    }
+  }
+};
+
+Pass::Status InlinePass::ProcessImpl() {
+  // do exhaustive inlining on each entry point function in module
```

Comment:
Yes. I think this will be a separate pass. Currently spirv-remap does a lot of dead code elimination (including dead function), although eventually I think that capability should migrate into spirv-opt.